Catsbi 7주차 미션 제출합니다.

catsbi opened the pull request:
안녕하세요. 명절 이후 오랜만에 PR인 것 같습니다. ㅎ
 인강 내용을 토대로 스프링 시큐리티를 적용한 뒤, 이전 주차 및 복습과정에서 작성했던 코드스타일로 전부는 아니지만 조금씩 리팩토링을 진행하고 있습니다. 

### 질의사항
1. 현재 조회 API및 서비스의 반환타입이 List<T> 인데 이전에 소트웍스 앤솔러지의 객체지향생활체조 원칙 중 일급 컬렉션을 사용하라는 말이 있어서 이를 ProductList, UserList등의 일급컬렉션으로 만들어서 반환을 하도록 하는건 어떨까요?  더하여 서비스는 도메인 목록을 가진 일급 컬렉션을 반환할테고 화면단에 노출될 View 목록으로 변환해주는 작업은 일급 컬렉션에서 담당하도록 하는 것이죠.  

2. @Valid 같은 어노테이션 기반 검증기는 MockMvc나 RestAssured같은 Rest API 테스트시에는 정상적으로 동작을 하니 문제가 안되지만, 도메인 혹은 서비스 테스트를 할 경우에는 해당 어노테이션이 동작을 하지않아 실패테스트를 작성하기가 어렵더라구요.  그래서 어찌해야하나 고민하다가 구글링 해보니 검증기를 임의로 가져와서 validate()메서드를 수행해서 나오는 Set 집합을 토대로 판단을 하던데, 이 방법으로 충분한가요? 

3. 현재 test 를 npm run test로 구동하면 patch 테스트와 delete 테스트의 결과 expected값이 상이합니다. 
둘 다 UserNotFoundException인데 하나는 403상태코드를 예측하고 한 쪽은 404 상태코드를 예측합니다. 

같은 예외인데 다른 상태코드를 변환하도록 하라는 의도가 맞나요?

Commit:
feat: Spring Security 학습 내용 적용

Commit:
feat: 암호화 학습내용 적용, h2-console 접근 차단 해제 설정 추가

Commit:
feat: lombok 애노테이션 메서드 코드 커버리지 exclude 설정 추가

catsbi committed:
feat: 식별자 및 도메인 Supplier 생성

catsbi committed:
feat: Supplier 구현체에 적용 및 정의 메소드 구현, 테스트 케이스 수정

hannut91 commented:
> 현재 test 를 npm run test로 구동하면 patch 테스트와 delete 테스트의 결과 expected값이 상이합니다.
둘 다 UserNotFoundException인데 하나는 403상태코드를 예측하고 한 쪽은 404 상태코드를 예측합니다.

403을 반환하도록한 이유는 권한이 없는 경우를 다르게 반환하게 할려고 했습니다 ㅎㅎ 테스트가 의도가 드러나지 않았나보군요 😢

hannut91 commented:
> 현재 조회 API및 서비스의 반환타입이 List 인데 이전에 소트웍스 앤솔러지의 객체지향생활체조 원칙 중 일급 컬렉션을 사용하라는 말이 있어서 이를 ProductList, UserList등의 일급컬렉션으로 만들어서 반환을 하도록 하는건 어떨까요? 더하여 서비스는 도메인 목록을 가진 일급 컬렉션을 반환할테고 화면단에 노출될 View 목록으로 변환해주는 작업은 일급 컬렉션에서 담당하도록 하는 것이죠.

저도 UserList와 ProductList를 객체로 만드는 것에는 동의하지만, 언제 유용한지 잘 떠오르지 않네요. 예제를 들어주실 수 있나요?

hannut91 commented:
> `@Valid` 같은 어노테이션 기반 검증기는 MockMvc나 RestAssured같은 Rest API 테스트시에는 정상적으로 동작을 하니 문제가 안되지만, 도메인 혹은 서비스 테스트를 할 경우에는 해당 어노테이션이 동작을 하지않아 실패테스트를 작성하기가 어렵더라구요. 그래서 어찌해야하나 고민하다가 구글링 해보니 검증기를 임의로 가져와서 validate()메서드를 수행해서 나오는 Set 집합을 토대로 판단을 하던데, 이 방법으로 충분한가요?

 음 이것도 어떤건지 코드로 보여주시면 좋을 것 같아요

catsbi committed:
feat: 사용자 정의 예외 객체 생성

catsbi committed:
feat: jwtUtil 수정, 검증기 유틸 클래스 생성, 상품 목록 일급 컬렉션과 ViewSupplier 생성, 예외 핸들러 추가

catsbi committed:
fix: 수정된 내용(일급 컬렉션, 인가 검증) 적용

catsbi committed:
refactor: reformat code

catsbi commented:
> 403을 반환하도록한 이유는 권한이 없는 경우를 다르게 반환하게 할려고 했습니다 ㅎㅎ 테스트가 의도가 드러나지 않았나보군요 😢
: 아, 둘 다 유저아이디를 못찾는 상황이라고 생각해 동일한 예왼데 다르게 반환해야하는게 이해가 안됐는데, 테스트쪽을 좀 더 살펴보니 한 쪽은 다른 아이디로 변경을 시도하고 다른 계정의 정보를 수정하려는건 인가되지 않기에 403을 반환하려는거였군요. 
새로운 예외객체를 만들어 처리해줬습니다. 

> 저도 UserList와 ProductList를 객체로 만드는 것에는 동의하지만, 언제 유용한지 잘 떠오르지 않네요. 예제를 들어주실 수 있나요?
: 현재 ProductList라고 ProductController의 조회부분 API에 구현했습니다. 

> 음 이것도 어떤건지 코드로 보여주시면 좋을 것 같아요
이것도 ProductService에 validOrThrow라는 메서드를 만들어서 구현해뒀습니다! 
```
    private void validOrThrow(EntitySupplier<Product> supplier) {
        final List<String> validateResult = Validators.validate(supplier);

        if (!validateResult.isEmpty()) {
            throw new InvalidProductArgumentsException(String.join(",", validateResult));
        }
    }
```

catsbi commented:
그 밖에는 퇴사를 앞두고 마지막으로 맡은 일이 좀 많아져서 회사에선 짬을 내지 못하니 퇴근하고 바로 미션을 작성하는데, 시간이 부족해서 리팩토링이 생각보다 많이 진행하진 못한 것 같습니다. 위에서 리뷰로 말해주신 부분들위주로 진행하고 그밖에 복습하면서 개선했던 내용들을 추가해봤습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -3,34 +3,40 @@
     }
 
-    public String login(String email, String password) {
+    public String login(Identifier identifier) {
```
`identifier`라고하니 식별자가 바로 떠오르네요. 당장 대안으로 생각난 것은 loginForm인데, 너무 프론트 관점인 것 같고...
로그인이 영어로 sign in이니까 signInData... 이것도 탈락인 것 같고.

시그니처만 다시보자면 

```
public String login(???)
```
반환하는 것이 문자열... 여기선 토큰이니까 토큰이라고 바꿔보면

```
public Token login(???)
```

저 물음표에 뭐가 들어가야 가장 자연스러울까 흠.. passport? 아닌 것 같고 

passport, loginData, loginForm, authenticationData

흠 그냥 email password를 냅두는게 좋을까요? 만약 email password로 로그인하지 않는 경우가 생기면 어쩌죠?

catsbi replied:
그냥 로그인에 필요한 식별정보기에 `Identifier`로 지정했는데, 식별자는 비밀번호는 포함하지 않는건가보네요. 
그럼 `AuthenticationDataSupplier` 정도로 만드는건 어떨까요? 시큐리티에는 이렇게 식별자와 인증키(비밀번호)를 제공하는 객체로 `UsernamePasswordAuthenticationToken` 를 제공하는데, 비슷하게 생각해봤습니다. 해당 객체와는 다르게 인터페이스를 통해 서비스를 제공하니 supplier를 덧붙혀서요 ㅎㅎ

hannut91 replied:
네 ㅎㅎ 좋습니다. 제가 궁금했던 건 이메일 비밀번호가 아닌 카카오톡 로그인이나 페이스북 로그인 같은 것으로 로그인을 할 때 어떻게 될지 궁금했습니다

hannut91 commented on app/src/test/java/com/codesoom/assignment/utils/JwtUtilTest.java:
```diff
@@ -32,9 +31,9 @@ void encode() {
 
     @Test
     void decodeWithValidToken() {
-        Claims claims = jwtUtil.decode(VALID_TOKEN);
+        final Long decodedId = jwtUtil.decode(VALID_TOKEN);
 
-        assertThat(claims.get("userId", Long.class)).isEqualTo(1L);
+        assertThat(decodedId).isEqualTo(1L);
     }
```
decodedId가 왜 1L인지 이해하기 어려운 것 같아요.

```java
String token = jwtUtil.encode(1L);

final Long id = jwtUtil.decode(token);

assertThat(id).isEqualTo(1L);
```

이렇게 쓰는 것은 어떨까요?

hannut91 commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -46,7 +67,7 @@ void authenticateWithDeletedUser() {
-        assertThat(user.authenticate("xxx")).isFalse();
+        assertThat(user.authenticate("test", passwordEncoder)).isFalse();
+        assertThat(user.authenticate("xxx", passwordEncoder)).isFalse();
     }
```
여기서 테스트가 말하고자 하는 것이 정확히 무엇인지 파악하기 어려운 것 같아요. 이렇게 아이디를 삭제했더니 인증에 실패했다고 보여주면 좋을 것 같아요

```java
User user = User.builder().build();
user.changePassword("test", passwordEncoder);

assertThat(user.authenticate("test", passwordEncoder)).isTrue();
assertThat(user.authenticate("xxx", passwordEncoder)).isFalse();

user.destroy();

assertThat(user.authenticate("test", passwordEncoder)).isFalse();
assertThat(user.authenticate("xxx", passwordEncoder)).isFalse();
```

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -14,21 +15,23 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 @WebMvcTest(ProductController.class)
 class ProductControllerTest {
-    private static final String VALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9." +
-            "eyJ1c2VySWQiOjF9.ZZ3CUl0jxeLGvQ1Js5nG2Ty5qGTlqai5ubDMXZOdaDk";
+    private static final String VALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOjF9.ZZ3CUl0jxeLGvQ1Js5nG2Ty5qGTlqai5ubDMXZOdaDk";
```
저는 가로 80줄로 제한해서 넘지 않도록 하는 편입니다.

catsbi replied:
postman 으로 테스트한다고 잠깐 줄 바꿔서 복사해놓고 돌려놓지를 않았네요 🥲

Commit:
refactor: 인터페이스 명 변경(Identifier -> AuthenticationDataSupplier) 외 코드 리팩터링

- DozerMapper 제거 및 EntitySupplier 인터페이스 활용
- JwtUtilTest, 기타 테스트 케이스 리뷰내용 적용
- Service Layer 메서드 매개변수 타입 EntitySupplier로 변경

Commit:
refactor: DozerMapper 빈 등록 메서드 제거, UserControllerTest 수정

catsbi committed:
docs: 주석 작성

catsbi committed:
refactor: 코드 리팩토링, ProductList 일급 컬렉션 테스트 케이스 작성, 테스트 불필요 메서드 exlucde추가

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -1,50 +1,63 @@
     }
 
+    private void validOrThrow(EntitySupplier<Product> supplier) {
+        final List<String> validateResult = Validators.validate(supplier);
```
반환타입이 리스트로 여러개를 나타내니, results같은 이름이 좋을 것 같아요. 그래야 isEmpty를 사용하는 것도 자연스럽게 보일 것 같습니다

Commit:
refactor: validate 메서드명 변경

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,19 +1,21 @@
 
 @SpringBootApplication
 public class App {
+    @Generated
     public static void main(String[] args) {
         SpringApplication.run(App.class, args);
     }
```
테스트 커버리지에서 제외할려고 하셨군요 ㅎㅎ 좋습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -1,50 +1,63 @@
     }
 
+    private void checkValidOrThrow(EntitySupplier<Product> supplier) {
+        final List<String> validateResult = Validators.getValidateResults(supplier);
```
이 Validation 결과도 ValidationResult라는 객체를 만들어서, 객체가 메시지를 만드는 일을 하도록 해야 할까요?

catsbi replied:
검증기 자체에서 완성된 메세지를 보내줘도 되겠군요

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -4,63 +4,102 @@
+    public List<ProductData> list() {
+        final ProductList products = productService.getProducts();
+        return products.convert(ProductData::from);
     }
```
저는 함수형 프로그래밍을 선호하다보니, 아직 이런 모양이 익숙한 것 같아요. `map`이라는 함수가 이미 무언가를 매핑하여 그대로 다시 List로 반환한다는 것을 알고 있어서 바로 이해할 수 있는데, `convert`는 이게 무엇으로 변경하는 것이지? 한 번 더 생각하게 되었던 것 같아요.

```java
return productService.getProducts()
    .map(ProductData::from);
```

catsbi replied:
컨버팅의 역할을 한다 생각해서 convert로 지었는데 생각해보니 어떤식으로 변환할지를 제공해주는 이상 좀 이상하긴하네요. map이 더 적절한 표현이 맞는것 같습니다. 수정하도록 하겠습니다.

catsbi committed:
refactor: convert 메서드명 변경

catsbi committed:
refactor: Validators에서 예외 메세지를 만드는 역할도 맡도록 수정

catsbi commented:
퇴사를 한 달 반 정도 남겨두고 갑자기 한달이면 된다며 프로젝트에 투입되면서 일이 너무 많아져서 어제는 작업을 하지 못했습니다 ㅠ 리뷰해주신 내용위주로 수정하여 PR드립니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/utils/JwtUtilTest.java:
```diff
@@ -23,18 +24,21 @@ void setUp() {
-        assertThat(claims.get("userId", Long.class)).isEqualTo(1L);
+    @ParameterizedTest
+    @NullSource
+    @ValueSource(longs = {Long.MIN_VALUE, -1, -30, -123901238, -87827273, -99999})
```
임계값을 테스트해주셨군요 ㅎㅎ 아주 좋은 습관입니다. 항상 어떤 경게에서 문제가 생기죠.

그런데 예외를 던지는 것을 보면 `NotSupportedUserIdException`으로 jwtUtil에서 사용자의 아이디 형식을 확인하고 있는 것 같아서 의문이 들었습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/domain/ProductListTest.java:
```diff
@@ -0,0 +1,201 @@
+        assertThat(productList.isEmpty()).isTrue();
+    }
+
+    @DisplayName("존재하지 않는 상품을 삭제요청해도 에러가 발생하지 않는다.")
+    @Test
+    void removeWithNotExistsProduct() {
+        final ProductList productList = ProductList.from(products);
+
+        final Product notExistsProduct = Product.builder()
+                .name("테스트")
+                .build();
+
+        assertThatCode(() -> productList.remove(notExistsProduct))
+                .doesNotThrowAnyException();
+    }
```
존재하는 상품을 삭제 요청하더라도 예외를 던지지 않도록 만드신 것 같네요. 존재하지 않는 상품을 삭제하면, 어떤 변화가 일어나는지, 혹은 일어나지 않음을 보여주면 좋을 것 같아요.

단순하게는 개수가 변하지 않음을 보여줄 수 있겠죠

catsbi replied:
그렇겠네요. 이런식으로 예외가 발생하지 않음을 테스트하는데 있어서 아직 어려움이 좀 있어서 고민이였는데, 사이즈를 이전과 이후로 비교하는방법이 있었네요

catsbi committed:
refactor: ProductListTest 삭제 테스트에 테스트 조건식 추가

catsbi commented:
백신 후유증인지 먹고있는 약 때문인지 여러모로 컨디션 난조로 피드백 적용이 좀 느렸습니다 ㅠㅠ
회사에서 몰래몰래 회고록 작성하고 퇴근해서 집도착하자마자 피드백 적용하고 다시 PR드립니다!
